### ISSUE TITLE
Infinite recursive processes started with version 1.1.242

### ISSUE THREAD

**Issue by jcrist** [RobertCraigie/pyright-python#60]
After upgrading pyright to 1.1.242, I'm seeing the following behavior. This persists after doing a clean install, and isn't seen with version 1.1.241:

```
$ PYRIGHT_PYTHON_DEBUG=1 pyright --help
2022-04-30 22:58:51,642 - DEBUG - pyright.utils - Latest pyright-python version is: 1.1.242
2022-04-30 22:58:51,642 - DEBUG - pyright.node - Checking for global target binary: npx
2022-04-30 22:58:51,642 - DEBUG - pyright.node - Global target binary: npx not found
2022-04-30 22:58:51,643 - DEBUG - pyright.node - Checking for nodeenv npx binary
2022-04-30 22:58:51,643 - DEBUG - pyright.node - Environment not found at /tmp/pyright-python.jcristharif/env
2022-04-30 22:58:51,643 - DEBUG - pyright.node - Installing nodeenv to /tmp/pyright-python.jcristharif/env
2022-04-30 22:58:51,643 - DEBUG - pyright.node - Running command with args: ['/home/jcristharif/miniconda3/envs/python-temp/bin/python3.9', '-m', 'nodeenv', '/tmp/pyright-python.jcristharif/env']
 * Install prebuilt node (18.0.0) ..... done.
2022-04-30 22:58:56,092 - DEBUG - pyright.node - Running node command with args: ['bash', '-c', '. /tmp/pyright-python.jcristharif/env/bin/activate && npx --version']
2022-04-30 22:58:56,298 - DEBUG - pyright.node - Version check for npx returning (8, 6, 0)
2022-04-30 22:57:26,438 - DEBUG - pyright.node - Checking for global target binary: npx
2022-04-30 22:57:26,439 - DEBUG - pyright.node - Global target binary: npx not found
2022-04-30 22:57:26,439 - DEBUG - pyright.node - Checking for nodeenv npx binary
2022-04-30 22:57:26,439 - DEBUG - pyright.node - Environment exists at /tmp/pyright-python.jcristharif/env
2022-04-30 22:57:26,440 - DEBUG - pyright.node - Running node command with args: ['bash', '-c', '. /tmp/pyright-python.jcristharif/env/bin/activate && npx --version']
2022-04-30 22:57:26,677 - DEBUG - pyright.node - Version check for npx returning (8, 6, 0)
2022-04-30 22:57:26,677 - DEBUG - pyright.node - Checking for global target binary: npx
2022-04-30 22:57:26,677 - DEBUG - pyright.node - Global target binary: npx not found
2022-04-30 22:57:26,678 - DEBUG - pyright.node - Checking for nodeenv npx binary
2022-04-30 22:57:26,678 - DEBUG - pyright.node - Environment exists at /tmp/pyright-python.jcristharif/env
2022-04-30 22:57:26,678 - DEBUG - pyright.node - Running node command with args: ['bash', '-c', '. /tmp/pyright-python.jcristharif/env/bin/activate && npx --silent --yes -- pyright@1.1.242 --help']
2022-04-30 22:57:27,582 - DEBUG - pyright.utils - Latest pyright-python version is: 1.1.242
# this just repeats for forever.
```

As best as I can tell the subprocess command `['bash', '-c', '. /tmp/pyright-python.jcristharif/env/bin/activate && npx --silent --yes -- pyright@1.1.242 --help']` is kicking off the Python CLI again, recursing for forever. After a few seconds a `killall pyright` terminates 10s of processes.

I'm running Linux, with no `nodejs` on `PATH`. Version 1.1.241 works fine, it's only on 1.1.242 that I see the above behavior.

**Comment by RobertCraigie** [comment]
> As best as I can tell the subprocess command ['bash', '-c', '. /tmp/pyright-python.jcristharif/env/bin/activate && npx --silent --yes -- pyright@1.1.242 --help'] is kicking off the Python CLI again, recursing for forever.

From looking at the logs it looks like your assessment is correct although I have no idea *how* that would happen.

There have also been no code changes for `pyright-python` between version `1.1.241` and `1.1.242` and as far as I'm aware there hasn't been a change in `pyright` itself that could cause an error like this so it is even more baffling as to what is causing it.

Did you see any potentially useful stack traces after killing the processes?

**Comment by jcrist** [comment]
> Did you see any potentially useful stack traces after killing the processes?

Nope, from the python side all processes are just waiting on the same `subprocess.run` call. Nothing interesting there.

Knowing nothing about nodejs, could this be due to a change in how `npx` finds the underlying `pyright` executable (either a change in pyright or a change in npx)?

**Comment by RobertCraigie** [comment]
> Knowing nothing about nodejs, could this be due to a change in how npx finds the underlying pyright executable (either a change in pyright or a change in npx)?

The library we use to download node (`nodeenv`) has not been updated which would mean that the same node version is used in `v1.1.241` and `v1.1.242` so nothing should have changed on the node side either.

I'm not aware of any change to the underlying `pyright` executable either.

**Comment by jcrist** [comment]
I'm seeing this behavior (recursive process creation) again. Only change I can think of is that I'm on a slow WIFI network:

Logs:

```
$ PYRIGHT_PYTHON_DEBUG=true pyright --version
2022-08-09 09:12:00,883 - DEBUG - pyright.utils - Latest pyright-python version is: 1.1.266
WARNING: there is a new pyright version available (v1.1.266).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

2022-08-09 09:12:00,883 - DEBUG - pyright.node - Checking for global target binary: npx
2022-08-09 09:12:00,884 - DEBUG - pyright.node - Found global binary at: /usr/bin/npx
2022-08-09 09:12:00,884 - DEBUG - pyright.node - Global binary exists at: /usr/bin/npx
2022-08-09 09:12:00,885 - DEBUG - pyright.node - Running node command with args: ['/usr/bin/npx', '--version']
2022-08-09 09:12:01,116 - DEBUG - pyright.node - Version check for npx returning (7, 5, 2)
2022-08-09 09:12:01,116 - DEBUG - pyright.node - Checking for global target binary: npx
2022-08-09 09:12:01,117 - DEBUG - pyright.node - Found global binary at: /usr/bin/npx
2022-08-09 09:12:01,117 - DEBUG - pyright.node - Global binary exists at: /usr/bin/npx
2022-08-09 09:12:01,117 - DEBUG - pyright.node - Running node command with args: ['/usr/bin/npx', '--silent', '--yes', '--', 'pyright@1.1.242', '--version']
2022-08-09 09:12:03,494 - DEBUG - pyright.utils - Latest pyright-python version is: 1.1.266
WARNING: there is a new pyright version available (v1.1.266).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

2022-08-09 09:12:03,494 - DEBUG - pyright.node - Checking for global target binary: npx
2022-08-09 09:12:03,494 - DEBUG - pyright.node - Found global binary at: /usr/bin/npx
2022-08-09 09:12:03,495 - DEBUG - pyright.node - Global binary exists at: /usr/bin/npx
2022-08-09 09:12:03,495 - DEBUG - pyright.node - Running node command with args: ['/usr/bin/npx', '--version']
2022-08-09 09:12:03,707 - DEBUG - pyright.node - Version check for npx returning (7, 5, 2)
2022-08-09 09:12:03,707 - DEBUG - pyright.node - Checking for global target binary: npx
2022-08-09 09:12:03,708 - DEBUG - pyright.node - Found global binary at: /usr/bin/npx
2022-08-09 09:12:03,708 - DEBUG - pyright.node - Global binary exists at: /usr/bin/npx
2022-08-09 09:12:03,708 - DEBUG - pyright.node - Running node command with args: ['/usr/bin/npx', '--silent', '--yes', '--', 'pyright@1.1.242', '--version']
2022-08-09 09:12:05,192 - DEBUG - pyright.utils - Latest pyright-python version is: 1.1.266
WARNING: there is a new pyright version available (v1.1.266).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

2022-08-09 09:12:05,192 - DEBUG - pyright.node - Checking for global target binary: npx
2022-08-09 09:12:05,193 - DEBUG - pyright.node - Found global binary at: /usr/bin/npx
2022-08-09 09:12:05,193 - DEBUG - pyright.node - Global binary exists at: /usr/bin/npx
2022-08-09 09:12:05,193 - DEBUG - pyright.node - Running node command with args: ['/usr/bin/npx', '--version']
...
```

If I fully disable wifi, then I still get recursive processes, but with slightly different logs:

```
$ PYRIGHT_PYTHON_DEBUG=true pyright --version
2022-08-09 09:14:19,900 - DEBUG - pyright.utils - Encountered exception while fetching latest release: <class 'http.client.HTTPException'> - [Errno -3] Temporary failure in name resolution
2022-08-09 09:14:19,901 - DEBUG - pyright.node - Checking for global target binary: npx
2022-08-09 09:14:19,901 - DEBUG - pyright.node - Found global binary at: /usr/bin/npx
2022-08-09 09:14:19,901 - DEBUG - pyright.node - Global binary exists at: /usr/bin/npx
2022-08-09 09:14:19,901 - DEBUG - pyright.node - Running node command with args: ['/usr/bin/npx', '--version']
2022-08-09 09:14:20,098 - DEBUG - pyright.node - Version check for npx returning (7, 5, 2)
2022-08-09 09:14:20,098 - DEBUG - pyright.node - Checking for global target binary: npx
2022-08-09 09:14:20,098 - DEBUG - pyright.node - Found global binary at: /usr/bin/npx
2022-08-09 09:14:20,098 - DEBUG - pyright.node - Global binary exists at: /usr/bin/npx
2022-08-09 09:14:20,098 - DEBUG - pyright.node - Running node command with args: ['/usr/bin/npx', '--silent', '--yes', '--', 'pyright@1.1.242', '--version']
2022-08-09 09:14:20,502 - DEBUG - pyright.utils - Encountered exception while fetching latest release: <class 'http.client.HTTPException'> - [Errno -3] Temporary failure in name resolution
2022-08-09 09:14:20,502 - DEBUG - pyright.node - Checking for global target binary: npx
2022-08-09 09:14:20,502 - DEBUG - pyright.node - Found global binary at: /usr/bin/npx
2022-08-09 09:14:20,502 - DEBUG - pyright.node - Global binary exists at: /usr/bin/npx
2022-08-09 09:14:20,502 - DEBUG - pyright.node - Running node command with args: ['/usr/bin/npx', '--version']
2022-08-09 09:14:20,700 - DEBUG - pyright.node - Version check for npx returning (7, 5, 2)
2022-08-09 09:14:20,700 - DEBUG - pyright.node - Checking for global target binary: npx
2022-08-09 09:14:20,700 - DEBUG - pyright.node - Found global binary at: /usr/bin/npx
2022-08-09 09:14:20,700 - DEBUG - pyright.node - Global binary exists at: /usr/bin/npx
2022-08-09 09:14:20,700 - DEBUG - pyright.node - Running node command with args: ['/usr/bin/npx', '--silent', '--yes', '--', 'pyright@1.1.242', '--version']
...
```

Setting `PYRIGHT_PYTHON_FORCE_VERSION=latest` fixes both of these, but I'm a little confused what it does. I read this environment variable (and the README) as telling pyright to always update itself to the latest version, but here it seems to allow avoiding the version check altogether, since wifi is disabled. Ideally I never want `pyright-python` to make a network request once it's installed - dependency versions are managed through a different mechanism. Is `PYRIGHT_PYTHON_FORCE_VERSION=latest` the way to accomplish that?

**Comment by RobertCraigie** [comment]
I believe this should now be resolved, the setup for installing pyright and node has changed substantially since this report. Please let me know if you run into it again.

---

The next release will add [`nodejs-wheel`](https://pypi.org/project/nodejs-wheel-binaries/) as an alternative to `nodeenv` which should be more reliable.

You can try it out today with
```
python -m pip install 'git+https://github.com/RobertCraigie/pyright-python.git#egg=pyright[nodejs]'
```
When released, it will be
```
pip install pyright[nodejs]
```
